// @workspace/api-client-react/src/generated/api.ts
/**
 * Generated by orval v8.5.3 🍺
 * Do not edit manually.
 * Api
 * Speech Clinic API for children ages 2-5
 * OpenAPI spec version: 0.1.0
 */
import { useMutation, useQuery } from "@tanstack/react-query";
import type {
  MutationFunction,
  QueryFunction,
  QueryKey,
  UseMutationOptions,
  UseMutationResult,
  UseQueryOptions,
  UseQueryResult,
} from "@tanstack/react-query";

import type {
  AdminDashboard,
  Appointment,
  Child,
  ChildDashboard,
  CreateAppointmentBody,
  CreateChildBody,
  CreateExerciseBody,
  CreateProgressBody,
  CreateSessionBody,
  CreateSpecialistBody,
  CreateUserBody,
  Exercise,
  HealthStatus,
  ListAppointmentsParams,
  ListChildrenParams,
  ListExercisesParams,
  ListProgressParams,
  ListSessionsParams,
  LoginBody,
  LoginResponse,
  ParentDashboard,
  Progress,
  Session,
  Specialist,
  SpecialistDashboard,
  UpdateAppointmentBody,
  UpdateChildBody,
  UpdateExerciseBody,
  UpdateSessionBody,
  UpdateSpecialistBody,
  UpdateUserBody,
  User,
} from "./api.schemas";

import { customFetch } from "../custom-fetch";
import type { ErrorType, BodyType } from "../custom-fetch";

type AwaitedInput<T> = PromiseLike<T> | T;

type Awaited<O> = O extends AwaitedInput<infer T> ? T : never;

type SecondParameter<T extends (...args: never) => unknown> = Parameters<T>[1];

/**
 * @summary Health check
 */
export const getHealthCheckUrl = () => {
  return `/api/healthz`;
};

export const healthCheck = async (
  options?: RequestInit,
): Promise<HealthStatus> => {
  return customFetch<HealthStatus>(getHealthCheckUrl(), {
    ...options,
    method: "GET",
  });
};

export const getHealthCheckQueryKey = () => {
  return [`/api/healthz`] as const;
};

export const getHealthCheckQueryOptions = <
  TData = Awaited<ReturnType<typeof healthCheck>>,
  TError = ErrorType<unknown>,
>(options?: {
  query?: UseQueryOptions<
    Awaited<ReturnType<typeof healthCheck>>,
    TError,
    TData
  >;
  request?: SecondParameter<typeof customFetch>;
}) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getHealthCheckQueryKey();

  const queryFn: QueryFunction<Awaited<ReturnType<typeof healthCheck>>> = ({
    signal,
  }) => healthCheck({ signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof healthCheck>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type HealthCheckQueryResult = NonNullable<
  Awaited<ReturnType<typeof healthCheck>>
>;
export type HealthCheckQueryError = ErrorType<unknown>;

/**
 * @summary Health check
 */

export function useHealthCheck<
  TData = Awaited<ReturnType<typeof healthCheck>>,
  TError = ErrorType<unknown>,
>(options?: {
  query?: UseQueryOptions<
    Awaited<ReturnType<typeof healthCheck>>,
    TError,
    TData
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getHealthCheckQueryOptions(options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Login
 */
export const getLoginUrl = () => {
  return `/api/auth/login`;
};

export const login = async (
  loginBody: LoginBody,
  options?: RequestInit,
): Promise<LoginResponse> => {
  return customFetch<LoginResponse>(getLoginUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(loginBody),
  });
};

export const getLoginMutationOptions = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof login>>,
    TError,
    { data: BodyType<LoginBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof login>>,
  TError,
  { data: BodyType<LoginBody> },
  TContext
> => {
  const mutationKey = ["login"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof login>>,
    { data: BodyType<LoginBody> }
  > = (props) => {
    const { data } = props ?? {};

    return login(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type LoginMutationResult = NonNullable<
  Awaited<ReturnType<typeof login>>
>;
export type LoginMutationBody = BodyType<LoginBody>;
export type LoginMutationError = ErrorType<void>;

/**
 * @summary Login
 */
export const useLogin = <
  TError = ErrorType<void>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof login>>,
    TError,
    { data: BodyType<LoginBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof login>>,
  TError,
  { data: BodyType<LoginBody> },
  TContext
> => {
  return useMutation(getLoginMutationOptions(options));
};

/**
 * @summary Logout
 */
export const getLogoutUrl = () => {
  return `/api/auth/logout`;
};

export const logout = async (options?: RequestInit): Promise<void> => {
  return customFetch<void>(getLogoutUrl(), {
    ...options,
    method: "POST",
  });
};

export const getLogoutMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof logout>>,
    TError,
    void,
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof logout>>,
  TError,
  void,
  TContext
> => {
  const mutationKey = ["logout"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof logout>>,
    void
  > = () => {
    return logout(requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type LogoutMutationResult = NonNullable<
  Awaited<ReturnType<typeof logout>>
>;

export type LogoutMutationError = ErrorType<unknown>;

/**
 * @summary Logout
 */
export const useLogout = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof logout>>,
    TError,
    void,
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof logout>>,
  TError,
  void,
  TContext
> => {
  return useMutation(getLogoutMutationOptions(options));
};

/**
 * @summary Get current user
 */
export const getGetCurrentUserUrl = () => {
  return `/api/auth/me`;
};

export const getCurrentUser = async (options?: RequestInit): Promise<User> => {
  return customFetch<User>(getGetCurrentUserUrl(), {
    ...options,
    method: "GET",
  });
};

export const getGetCurrentUserQueryKey = () => {
  return [`/api/auth/me`] as const;
};

export const getGetCurrentUserQueryOptions = <
  TData = Awaited<ReturnType<typeof getCurrentUser>>,
  TError = ErrorType<void>,
>(options?: {
  query?: UseQueryOptions<
    Awaited<ReturnType<typeof getCurrentUser>>,
    TError,
    TData
  >;
  request?: SecondParameter<typeof customFetch>;
}) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetCurrentUserQueryKey();

  const queryFn: QueryFunction<Awaited<ReturnType<typeof getCurrentUser>>> = ({
    signal,
  }) => getCurrentUser({ signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof getCurrentUser>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type GetCurrentUserQueryResult = NonNullable<
  Awaited<ReturnType<typeof getCurrentUser>>
>;
export type GetCurrentUserQueryError = ErrorType<void>;

/**
 * @summary Get current user
 */

export function useGetCurrentUser<
  TData = Awaited<ReturnType<typeof getCurrentUser>>,
  TError = ErrorType<void>,
>(options?: {
  query?: UseQueryOptions<
    Awaited<ReturnType<typeof getCurrentUser>>,
    TError,
    TData
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetCurrentUserQueryOptions(options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary List all users (admin only)
 */
export const getListUsersUrl = () => {
  return `/api/users`;
};

export const listUsers = async (options?: RequestInit): Promise<User[]> => {
  return customFetch<User[]>(getListUsersUrl(), {
    ...options,
    method: "GET",
  });
};

export const getListUsersQueryKey = () => {
  return [`/api/users`] as const;
};

export const getListUsersQueryOptions = <
  TData = Awaited<ReturnType<typeof listUsers>>,
  TError = ErrorType<unknown>,
>(options?: {
  query?: UseQueryOptions<Awaited<ReturnType<typeof listUsers>>, TError, TData>;
  request?: SecondParameter<typeof customFetch>;
}) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getListUsersQueryKey();

  const queryFn: QueryFunction<Awaited<ReturnType<typeof listUsers>>> = ({
    signal,
  }) => listUsers({ signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof listUsers>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type ListUsersQueryResult = NonNullable<
  Awaited<ReturnType<typeof listUsers>>
>;
export type ListUsersQueryError = ErrorType<unknown>;

/**
 * @summary List all users (admin only)
 */

export function useListUsers<
  TData = Awaited<ReturnType<typeof listUsers>>,
  TError = ErrorType<unknown>,
>(options?: {
  query?: UseQueryOptions<Awaited<ReturnType<typeof listUsers>>, TError, TData>;
  request?: SecondParameter<typeof customFetch>;
}): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getListUsersQueryOptions(options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Create a user (admin only)
 */
export const getCreateUserUrl = () => {
  return `/api/users`;
};

export const createUser = async (
  createUserBody: CreateUserBody,
  options?: RequestInit,
): Promise<User> => {
  return customFetch<User>(getCreateUserUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(createUserBody),
  });
};

export const getCreateUserMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createUser>>,
    TError,
    { data: BodyType<CreateUserBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof createUser>>,
  TError,
  { data: BodyType<CreateUserBody> },
  TContext
> => {
  const mutationKey = ["createUser"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof createUser>>,
    { data: BodyType<CreateUserBody> }
  > = (props) => {
    const { data } = props ?? {};

    return createUser(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type CreateUserMutationResult = NonNullable<
  Awaited<ReturnType<typeof createUser>>
>;
export type CreateUserMutationBody = BodyType<CreateUserBody>;
export type CreateUserMutationError = ErrorType<unknown>;

/**
 * @summary Create a user (admin only)
 */
export const useCreateUser = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createUser>>,
    TError,
    { data: BodyType<CreateUserBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof createUser>>,
  TError,
  { data: BodyType<CreateUserBody> },
  TContext
> => {
  return useMutation(getCreateUserMutationOptions(options));
};

/**
 * @summary Get user by ID
 */
export const getGetUserUrl = (id: number) => {
  return `/api/users/${id}`;
};

export const getUser = async (
  id: number,
  options?: RequestInit,
): Promise<User> => {
  return customFetch<User>(getGetUserUrl(id), {
    ...options,
    method: "GET",
  });
};

export const getGetUserQueryKey = (id: number) => {
  return [`/api/users/${id}`] as const;
};

export const getGetUserQueryOptions = <
  TData = Awaited<ReturnType<typeof getUser>>,
  TError = ErrorType<void>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<Awaited<ReturnType<typeof getUser>>, TError, TData>;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetUserQueryKey(id);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof getUser>>> = ({
    signal,
  }) => getUser(id, { signal, ...requestOptions });

  return {
    queryKey,
    queryFn,
    enabled: !!id,
    ...queryOptions,
  } as UseQueryOptions<Awaited<ReturnType<typeof getUser>>, TError, TData> & {
    queryKey: QueryKey;
  };
};

export type GetUserQueryResult = NonNullable<
  Awaited<ReturnType<typeof getUser>>
>;
export type GetUserQueryError = ErrorType<void>;

/**
 * @summary Get user by ID
 */

export function useGetUser<
  TData = Awaited<ReturnType<typeof getUser>>,
  TError = ErrorType<void>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<Awaited<ReturnType<typeof getUser>>, TError, TData>;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetUserQueryOptions(id, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Update user
 */
export const getUpdateUserUrl = (id: number) => {
  return `/api/users/${id}`;
};

export const updateUser = async (
  id: number,
  updateUserBody: UpdateUserBody,
  options?: RequestInit,
): Promise<User> => {
  return customFetch<User>(getUpdateUserUrl(id), {
    ...options,
    method: "PATCH",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(updateUserBody),
  });
};

export const getUpdateUserMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateUser>>,
    TError,
    { id: number; data: BodyType<UpdateUserBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof updateUser>>,
  TError,
  { id: number; data: BodyType<UpdateUserBody> },
  TContext
> => {
  const mutationKey = ["updateUser"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof updateUser>>,
    { id: number; data: BodyType<UpdateUserBody> }
  > = (props) => {
    const { id, data } = props ?? {};

    return updateUser(id, data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type UpdateUserMutationResult = NonNullable<
  Awaited<ReturnType<typeof updateUser>>
>;
export type UpdateUserMutationBody = BodyType<UpdateUserBody>;
export type UpdateUserMutationError = ErrorType<unknown>;

/**
 * @summary Update user
 */
export const useUpdateUser = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateUser>>,
    TError,
    { id: number; data: BodyType<UpdateUserBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof updateUser>>,
  TError,
  { id: number; data: BodyType<UpdateUserBody> },
  TContext
> => {
  return useMutation(getUpdateUserMutationOptions(options));
};

/**
 * @summary Delete user
 */
export const getDeleteUserUrl = (id: number) => {
  return `/api/users/${id}`;
};

export const deleteUser = async (
  id: number,
  options?: RequestInit,
): Promise<void> => {
  return customFetch<void>(getDeleteUserUrl(id), {
    ...options,
    method: "DELETE",
  });
};

export const getDeleteUserMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof deleteUser>>,
    TError,
    { id: number },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof deleteUser>>,
  TError,
  { id: number },
  TContext
> => {
  const mutationKey = ["deleteUser"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof deleteUser>>,
    { id: number }
  > = (props) => {
    const { id } = props ?? {};

    return deleteUser(id, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type DeleteUserMutationResult = NonNullable<
  Awaited<ReturnType<typeof deleteUser>>
>;

export type DeleteUserMutationError = ErrorType<unknown>;

/**
 * @summary Delete user
 */
export const useDeleteUser = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof deleteUser>>,
    TError,
    { id: number },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof deleteUser>>,
  TError,
  { id: number },
  TContext
> => {
  return useMutation(getDeleteUserMutationOptions(options));
};

/**
 * @summary List all specialists
 */
export const getListSpecialistsUrl = () => {
  return `/api/specialists`;
};

export const listSpecialists = async (
  options?: RequestInit,
): Promise<Specialist[]> => {
  return customFetch<Specialist[]>(getListSpecialistsUrl(), {
    ...options,
    method: "GET",
  });
};

export const getListSpecialistsQueryKey = () => {
  return [`/api/specialists`] as const;
};

export const getListSpecialistsQueryOptions = <
  TData = Awaited<ReturnType<typeof listSpecialists>>,
  TError = ErrorType<unknown>,
>(options?: {
  query?: UseQueryOptions<
    Awaited<ReturnType<typeof listSpecialists>>,
    TError,
    TData
  >;
  request?: SecondParameter<typeof customFetch>;
}) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getListSpecialistsQueryKey();

  const queryFn: QueryFunction<Awaited<ReturnType<typeof listSpecialists>>> = ({
    signal,
  }) => listSpecialists({ signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof listSpecialists>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type ListSpecialistsQueryResult = NonNullable<
  Awaited<ReturnType<typeof listSpecialists>>
>;
export type ListSpecialistsQueryError = ErrorType<unknown>;

/**
 * @summary List all specialists
 */

export function useListSpecialists<
  TData = Awaited<ReturnType<typeof listSpecialists>>,
  TError = ErrorType<unknown>,
>(options?: {
  query?: UseQueryOptions<
    Awaited<ReturnType<typeof listSpecialists>>,
    TError,
    TData
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getListSpecialistsQueryOptions(options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Create specialist profile
 */
export const getCreateSpecialistUrl = () => {
  return `/api/specialists`;
};

export const createSpecialist = async (
  createSpecialistBody: CreateSpecialistBody,
  options?: RequestInit,
): Promise<Specialist> => {
  return customFetch<Specialist>(getCreateSpecialistUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(createSpecialistBody),
  });
};

export const getCreateSpecialistMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createSpecialist>>,
    TError,
    { data: BodyType<CreateSpecialistBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof createSpecialist>>,
  TError,
  { data: BodyType<CreateSpecialistBody> },
  TContext
> => {
  const mutationKey = ["createSpecialist"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof createSpecialist>>,
    { data: BodyType<CreateSpecialistBody> }
  > = (props) => {
    const { data } = props ?? {};

    return createSpecialist(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type CreateSpecialistMutationResult = NonNullable<
  Awaited<ReturnType<typeof createSpecialist>>
>;
export type CreateSpecialistMutationBody = BodyType<CreateSpecialistBody>;
export type CreateSpecialistMutationError = ErrorType<unknown>;

/**
 * @summary Create specialist profile
 */
export const useCreateSpecialist = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createSpecialist>>,
    TError,
    { data: BodyType<CreateSpecialistBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof createSpecialist>>,
  TError,
  { data: BodyType<CreateSpecialistBody> },
  TContext
> => {
  return useMutation(getCreateSpecialistMutationOptions(options));
};

/**
 * @summary Get specialist by ID
 */
export const getGetSpecialistUrl = (id: number) => {
  return `/api/specialists/${id}`;
};

export const getSpecialist = async (
  id: number,
  options?: RequestInit,
): Promise<Specialist> => {
  return customFetch<Specialist>(getGetSpecialistUrl(id), {
    ...options,
    method: "GET",
  });
};

export const getGetSpecialistQueryKey = (id: number) => {
  return [`/api/specialists/${id}`] as const;
};

export const getGetSpecialistQueryOptions = <
  TData = Awaited<ReturnType<typeof getSpecialist>>,
  TError = ErrorType<unknown>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getSpecialist>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetSpecialistQueryKey(id);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof getSpecialist>>> = ({
    signal,
  }) => getSpecialist(id, { signal, ...requestOptions });

  return {
    queryKey,
    queryFn,
    enabled: !!id,
    ...queryOptions,
  } as UseQueryOptions<
    Awaited<ReturnType<typeof getSpecialist>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type GetSpecialistQueryResult = NonNullable<
  Awaited<ReturnType<typeof getSpecialist>>
>;
export type GetSpecialistQueryError = ErrorType<unknown>;

/**
 * @summary Get specialist by ID
 */

export function useGetSpecialist<
  TData = Awaited<ReturnType<typeof getSpecialist>>,
  TError = ErrorType<unknown>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getSpecialist>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetSpecialistQueryOptions(id, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Update specialist
 */
export const getUpdateSpecialistUrl = (id: number) => {
  return `/api/specialists/${id}`;
};

export const updateSpecialist = async (
  id: number,
  updateSpecialistBody: UpdateSpecialistBody,
  options?: RequestInit,
): Promise<Specialist> => {
  return customFetch<Specialist>(getUpdateSpecialistUrl(id), {
    ...options,
    method: "PATCH",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(updateSpecialistBody),
  });
};

export const getUpdateSpecialistMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateSpecialist>>,
    TError,
    { id: number; data: BodyType<UpdateSpecialistBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof updateSpecialist>>,
  TError,
  { id: number; data: BodyType<UpdateSpecialistBody> },
  TContext
> => {
  const mutationKey = ["updateSpecialist"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof updateSpecialist>>,
    { id: number; data: BodyType<UpdateSpecialistBody> }
  > = (props) => {
    const { id, data } = props ?? {};

    return updateSpecialist(id, data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type UpdateSpecialistMutationResult = NonNullable<
  Awaited<ReturnType<typeof updateSpecialist>>
>;
export type UpdateSpecialistMutationBody = BodyType<UpdateSpecialistBody>;
export type UpdateSpecialistMutationError = ErrorType<unknown>;

/**
 * @summary Update specialist
 */
export const useUpdateSpecialist = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateSpecialist>>,
    TError,
    { id: number; data: BodyType<UpdateSpecialistBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof updateSpecialist>>,
  TError,
  { id: number; data: BodyType<UpdateSpecialistBody> },
  TContext
> => {
  return useMutation(getUpdateSpecialistMutationOptions(options));
};

/**
 * @summary List children
 */
export const getListChildrenUrl = (params?: ListChildrenParams) => {
  const normalizedParams = new URLSearchParams();

  Object.entries(params || {}).forEach(([key, value]) => {
    if (value !== undefined) {
      normalizedParams.append(key, value === null ? "null" : value.toString());
    }
  });

  const stringifiedParams = normalizedParams.toString();

  return stringifiedParams.length > 0
    ? `/api/children?${stringifiedParams}`
    : `/api/children`;
};

export const listChildren = async (
  params?: ListChildrenParams,
  options?: RequestInit,
): Promise<Child[]> => {
  return customFetch<Child[]>(getListChildrenUrl(params), {
    ...options,
    method: "GET",
  });
};

export const getListChildrenQueryKey = (params?: ListChildrenParams) => {
  return [`/api/children`, ...(params ? [params] : [])] as const;
};

export const getListChildrenQueryOptions = <
  TData = Awaited<ReturnType<typeof listChildren>>,
  TError = ErrorType<unknown>,
>(
  params?: ListChildrenParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof listChildren>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getListChildrenQueryKey(params);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof listChildren>>> = ({
    signal,
  }) => listChildren(params, { signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof listChildren>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type ListChildrenQueryResult = NonNullable<
  Awaited<ReturnType<typeof listChildren>>
>;
export type ListChildrenQueryError = ErrorType<unknown>;

/**
 * @summary List children
 */

export function useListChildren<
  TData = Awaited<ReturnType<typeof listChildren>>,
  TError = ErrorType<unknown>,
>(
  params?: ListChildrenParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof listChildren>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getListChildrenQueryOptions(params, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Create child profile
 */
export const getCreateChildUrl = () => {
  return `/api/children`;
};

export const createChild = async (
  createChildBody: CreateChildBody,
  options?: RequestInit,
): Promise<Child> => {
  return customFetch<Child>(getCreateChildUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(createChildBody),
  });
};

export const getCreateChildMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createChild>>,
    TError,
    { data: BodyType<CreateChildBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof createChild>>,
  TError,
  { data: BodyType<CreateChildBody> },
  TContext
> => {
  const mutationKey = ["createChild"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof createChild>>,
    { data: BodyType<CreateChildBody> }
  > = (props) => {
    const { data } = props ?? {};

    return createChild(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type CreateChildMutationResult = NonNullable<
  Awaited<ReturnType<typeof createChild>>
>;
export type CreateChildMutationBody = BodyType<CreateChildBody>;
export type CreateChildMutationError = ErrorType<unknown>;

/**
 * @summary Create child profile
 */
export const useCreateChild = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createChild>>,
    TError,
    { data: BodyType<CreateChildBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof createChild>>,
  TError,
  { data: BodyType<CreateChildBody> },
  TContext
> => {
  return useMutation(getCreateChildMutationOptions(options));
};

/**
 * @summary Get child by ID
 */
export const getGetChildUrl = (id: number) => {
  return `/api/children/${id}`;
};

export const getChild = async (
  id: number,
  options?: RequestInit,
): Promise<Child> => {
  return customFetch<Child>(getGetChildUrl(id), {
    ...options,
    method: "GET",
  });
};

export const getGetChildQueryKey = (id: number) => {
  return [`/api/children/${id}`] as const;
};

export const getGetChildQueryOptions = <
  TData = Awaited<ReturnType<typeof getChild>>,
  TError = ErrorType<void>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getChild>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetChildQueryKey(id);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof getChild>>> = ({
    signal,
  }) => getChild(id, { signal, ...requestOptions });

  return {
    queryKey,
    queryFn,
    enabled: !!id,
    ...queryOptions,
  } as UseQueryOptions<Awaited<ReturnType<typeof getChild>>, TError, TData> & {
    queryKey: QueryKey;
  };
};

export type GetChildQueryResult = NonNullable<
  Awaited<ReturnType<typeof getChild>>
>;
export type GetChildQueryError = ErrorType<void>;

/**
 * @summary Get child by ID
 */

export function useGetChild<
  TData = Awaited<ReturnType<typeof getChild>>,
  TError = ErrorType<void>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getChild>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetChildQueryOptions(id, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Update child
 */
export const getUpdateChildUrl = (id: number) => {
  return `/api/children/${id}`;
};

export const updateChild = async (
  id: number,
  updateChildBody: UpdateChildBody,
  options?: RequestInit,
): Promise<Child> => {
  return customFetch<Child>(getUpdateChildUrl(id), {
    ...options,
    method: "PATCH",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(updateChildBody),
  });
};

export const getUpdateChildMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateChild>>,
    TError,
    { id: number; data: BodyType<UpdateChildBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof updateChild>>,
  TError,
  { id: number; data: BodyType<UpdateChildBody> },
  TContext
> => {
  const mutationKey = ["updateChild"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof updateChild>>,
    { id: number; data: BodyType<UpdateChildBody> }
  > = (props) => {
    const { id, data } = props ?? {};

    return updateChild(id, data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type UpdateChildMutationResult = NonNullable<
  Awaited<ReturnType<typeof updateChild>>
>;
export type UpdateChildMutationBody = BodyType<UpdateChildBody>;
export type UpdateChildMutationError = ErrorType<unknown>;

/**
 * @summary Update child
 */
export const useUpdateChild = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateChild>>,
    TError,
    { id: number; data: BodyType<UpdateChildBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof updateChild>>,
  TError,
  { id: number; data: BodyType<UpdateChildBody> },
  TContext
> => {
  return useMutation(getUpdateChildMutationOptions(options));
};

/**
 * @summary Delete child
 */
export const getDeleteChildUrl = (id: number) => {
  return `/api/children/${id}`;
};

export const deleteChild = async (
  id: number,
  options?: RequestInit,
): Promise<void> => {
  return customFetch<void>(getDeleteChildUrl(id), {
    ...options,
    method: "DELETE",
  });
};

export const getDeleteChildMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof deleteChild>>,
    TError,
    { id: number },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof deleteChild>>,
  TError,
  { id: number },
  TContext
> => {
  const mutationKey = ["deleteChild"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof deleteChild>>,
    { id: number }
  > = (props) => {
    const { id } = props ?? {};

    return deleteChild(id, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type DeleteChildMutationResult = NonNullable<
  Awaited<ReturnType<typeof deleteChild>>
>;

export type DeleteChildMutationError = ErrorType<unknown>;

/**
 * @summary Delete child
 */
export const useDeleteChild = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof deleteChild>>,
    TError,
    { id: number },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof deleteChild>>,
  TError,
  { id: number },
  TContext
> => {
  return useMutation(getDeleteChildMutationOptions(options));
};

/**
 * @summary List appointments
 */
export const getListAppointmentsUrl = (params?: ListAppointmentsParams) => {
  const normalizedParams = new URLSearchParams();

  Object.entries(params || {}).forEach(([key, value]) => {
    if (value !== undefined) {
      normalizedParams.append(key, value === null ? "null" : value.toString());
    }
  });

  const stringifiedParams = normalizedParams.toString();

  return stringifiedParams.length > 0
    ? `/api/appointments?${stringifiedParams}`
    : `/api/appointments`;
};

export const listAppointments = async (
  params?: ListAppointmentsParams,
  options?: RequestInit,
): Promise<Appointment[]> => {
  return customFetch<Appointment[]>(getListAppointmentsUrl(params), {
    ...options,
    method: "GET",
  });
};

export const getListAppointmentsQueryKey = (
  params?: ListAppointmentsParams,
) => {
  return [`/api/appointments`, ...(params ? [params] : [])] as const;
};

export const getListAppointmentsQueryOptions = <
  TData = Awaited<ReturnType<typeof listAppointments>>,
  TError = ErrorType<unknown>,
>(
  params?: ListAppointmentsParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof listAppointments>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey =
    queryOptions?.queryKey ?? getListAppointmentsQueryKey(params);

  const queryFn: QueryFunction<
    Awaited<ReturnType<typeof listAppointments>>
  > = ({ signal }) => listAppointments(params, { signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof listAppointments>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type ListAppointmentsQueryResult = NonNullable<
  Awaited<ReturnType<typeof listAppointments>>
>;
export type ListAppointmentsQueryError = ErrorType<unknown>;

/**
 * @summary List appointments
 */

export function useListAppointments<
  TData = Awaited<ReturnType<typeof listAppointments>>,
  TError = ErrorType<unknown>,
>(
  params?: ListAppointmentsParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof listAppointments>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getListAppointmentsQueryOptions(params, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Create appointment
 */
export const getCreateAppointmentUrl = () => {
  return `/api/appointments`;
};

export const createAppointment = async (
  createAppointmentBody: CreateAppointmentBody,
  options?: RequestInit,
): Promise<Appointment> => {
  return customFetch<Appointment>(getCreateAppointmentUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(createAppointmentBody),
  });
};

export const getCreateAppointmentMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createAppointment>>,
    TError,
    { data: BodyType<CreateAppointmentBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof createAppointment>>,
  TError,
  { data: BodyType<CreateAppointmentBody> },
  TContext
> => {
  const mutationKey = ["createAppointment"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof createAppointment>>,
    { data: BodyType<CreateAppointmentBody> }
  > = (props) => {
    const { data } = props ?? {};

    return createAppointment(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type CreateAppointmentMutationResult = NonNullable<
  Awaited<ReturnType<typeof createAppointment>>
>;
export type CreateAppointmentMutationBody = BodyType<CreateAppointmentBody>;
export type CreateAppointmentMutationError = ErrorType<unknown>;

/**
 * @summary Create appointment
 */
export const useCreateAppointment = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createAppointment>>,
    TError,
    { data: BodyType<CreateAppointmentBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof createAppointment>>,
  TError,
  { data: BodyType<CreateAppointmentBody> },
  TContext
> => {
  return useMutation(getCreateAppointmentMutationOptions(options));
};

/**
 * @summary Get appointment by ID
 */
export const getGetAppointmentUrl = (id: number) => {
  return `/api/appointments/${id}`;
};

export const getAppointment = async (
  id: number,
  options?: RequestInit,
): Promise<Appointment> => {
  return customFetch<Appointment>(getGetAppointmentUrl(id), {
    ...options,
    method: "GET",
  });
};

export const getGetAppointmentQueryKey = (id: number) => {
  return [`/api/appointments/${id}`] as const;
};

export const getGetAppointmentQueryOptions = <
  TData = Awaited<ReturnType<typeof getAppointment>>,
  TError = ErrorType<unknown>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getAppointment>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetAppointmentQueryKey(id);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof getAppointment>>> = ({
    signal,
  }) => getAppointment(id, { signal, ...requestOptions });

  return {
    queryKey,
    queryFn,
    enabled: !!id,
    ...queryOptions,
  } as UseQueryOptions<
    Awaited<ReturnType<typeof getAppointment>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type GetAppointmentQueryResult = NonNullable<
  Awaited<ReturnType<typeof getAppointment>>
>;
export type GetAppointmentQueryError = ErrorType<unknown>;

/**
 * @summary Get appointment by ID
 */

export function useGetAppointment<
  TData = Awaited<ReturnType<typeof getAppointment>>,
  TError = ErrorType<unknown>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getAppointment>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetAppointmentQueryOptions(id, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Update appointment
 */
export const getUpdateAppointmentUrl = (id: number) => {
  return `/api/appointments/${id}`;
};

export const updateAppointment = async (
  id: number,
  updateAppointmentBody: UpdateAppointmentBody,
  options?: RequestInit,
): Promise<Appointment> => {
  return customFetch<Appointment>(getUpdateAppointmentUrl(id), {
    ...options,
    method: "PATCH",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(updateAppointmentBody),
  });
};

export const getUpdateAppointmentMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateAppointment>>,
    TError,
    { id: number; data: BodyType<UpdateAppointmentBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof updateAppointment>>,
  TError,
  { id: number; data: BodyType<UpdateAppointmentBody> },
  TContext
> => {
  const mutationKey = ["updateAppointment"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof updateAppointment>>,
    { id: number; data: BodyType<UpdateAppointmentBody> }
  > = (props) => {
    const { id, data } = props ?? {};

    return updateAppointment(id, data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type UpdateAppointmentMutationResult = NonNullable<
  Awaited<ReturnType<typeof updateAppointment>>
>;
export type UpdateAppointmentMutationBody = BodyType<UpdateAppointmentBody>;
export type UpdateAppointmentMutationError = ErrorType<unknown>;

/**
 * @summary Update appointment
 */
export const useUpdateAppointment = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateAppointment>>,
    TError,
    { id: number; data: BodyType<UpdateAppointmentBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof updateAppointment>>,
  TError,
  { id: number; data: BodyType<UpdateAppointmentBody> },
  TContext
> => {
  return useMutation(getUpdateAppointmentMutationOptions(options));
};

/**
 * @summary Delete appointment
 */
export const getDeleteAppointmentUrl = (id: number) => {
  return `/api/appointments/${id}`;
};

export const deleteAppointment = async (
  id: number,
  options?: RequestInit,
): Promise<void> => {
  return customFetch<void>(getDeleteAppointmentUrl(id), {
    ...options,
    method: "DELETE",
  });
};

export const getDeleteAppointmentMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof deleteAppointment>>,
    TError,
    { id: number },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof deleteAppointment>>,
  TError,
  { id: number },
  TContext
> => {
  const mutationKey = ["deleteAppointment"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof deleteAppointment>>,
    { id: number }
  > = (props) => {
    const { id } = props ?? {};

    return deleteAppointment(id, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type DeleteAppointmentMutationResult = NonNullable<
  Awaited<ReturnType<typeof deleteAppointment>>
>;

export type DeleteAppointmentMutationError = ErrorType<unknown>;

/**
 * @summary Delete appointment
 */
export const useDeleteAppointment = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof deleteAppointment>>,
    TError,
    { id: number },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof deleteAppointment>>,
  TError,
  { id: number },
  TContext
> => {
  return useMutation(getDeleteAppointmentMutationOptions(options));
};

/**
 * @summary List therapy sessions
 */
export const getListSessionsUrl = (params?: ListSessionsParams) => {
  const normalizedParams = new URLSearchParams();

  Object.entries(params || {}).forEach(([key, value]) => {
    if (value !== undefined) {
      normalizedParams.append(key, value === null ? "null" : value.toString());
    }
  });

  const stringifiedParams = normalizedParams.toString();

  return stringifiedParams.length > 0
    ? `/api/sessions?${stringifiedParams}`
    : `/api/sessions`;
};

export const listSessions = async (
  params?: ListSessionsParams,
  options?: RequestInit,
): Promise<Session[]> => {
  return customFetch<Session[]>(getListSessionsUrl(params), {
    ...options,
    method: "GET",
  });
};

export const getListSessionsQueryKey = (params?: ListSessionsParams) => {
  return [`/api/sessions`, ...(params ? [params] : [])] as const;
};

export const getListSessionsQueryOptions = <
  TData = Awaited<ReturnType<typeof listSessions>>,
  TError = ErrorType<unknown>,
>(
  params?: ListSessionsParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof listSessions>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getListSessionsQueryKey(params);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof listSessions>>> = ({
    signal,
  }) => listSessions(params, { signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof listSessions>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type ListSessionsQueryResult = NonNullable<
  Awaited<ReturnType<typeof listSessions>>
>;
export type ListSessionsQueryError = ErrorType<unknown>;

/**
 * @summary List therapy sessions
 */

export function useListSessions<
  TData = Awaited<ReturnType<typeof listSessions>>,
  TError = ErrorType<unknown>,
>(
  params?: ListSessionsParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof listSessions>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getListSessionsQueryOptions(params, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Create therapy session
 */
export const getCreateSessionUrl = () => {
  return `/api/sessions`;
};

export const createSession = async (
  createSessionBody: CreateSessionBody,
  options?: RequestInit,
): Promise<Session> => {
  return customFetch<Session>(getCreateSessionUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(createSessionBody),
  });
};

export const getCreateSessionMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createSession>>,
    TError,
    { data: BodyType<CreateSessionBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof createSession>>,
  TError,
  { data: BodyType<CreateSessionBody> },
  TContext
> => {
  const mutationKey = ["createSession"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof createSession>>,
    { data: BodyType<CreateSessionBody> }
  > = (props) => {
    const { data } = props ?? {};

    return createSession(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type CreateSessionMutationResult = NonNullable<
  Awaited<ReturnType<typeof createSession>>
>;
export type CreateSessionMutationBody = BodyType<CreateSessionBody>;
export type CreateSessionMutationError = ErrorType<unknown>;

/**
 * @summary Create therapy session
 */
export const useCreateSession = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createSession>>,
    TError,
    { data: BodyType<CreateSessionBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof createSession>>,
  TError,
  { data: BodyType<CreateSessionBody> },
  TContext
> => {
  return useMutation(getCreateSessionMutationOptions(options));
};

/**
 * @summary Get session by ID
 */
export const getGetSessionUrl = (id: number) => {
  return `/api/sessions/${id}`;
};

export const getSession = async (
  id: number,
  options?: RequestInit,
): Promise<Session> => {
  return customFetch<Session>(getGetSessionUrl(id), {
    ...options,
    method: "GET",
  });
};

export const getGetSessionQueryKey = (id: number) => {
  return [`/api/sessions/${id}`] as const;
};

export const getGetSessionQueryOptions = <
  TData = Awaited<ReturnType<typeof getSession>>,
  TError = ErrorType<unknown>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getSession>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetSessionQueryKey(id);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof getSession>>> = ({
    signal,
  }) => getSession(id, { signal, ...requestOptions });

  return {
    queryKey,
    queryFn,
    enabled: !!id,
    ...queryOptions,
  } as UseQueryOptions<
    Awaited<ReturnType<typeof getSession>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type GetSessionQueryResult = NonNullable<
  Awaited<ReturnType<typeof getSession>>
>;
export type GetSessionQueryError = ErrorType<unknown>;

/**
 * @summary Get session by ID
 */

export function useGetSession<
  TData = Awaited<ReturnType<typeof getSession>>,
  TError = ErrorType<unknown>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getSession>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetSessionQueryOptions(id, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Update session
 */
export const getUpdateSessionUrl = (id: number) => {
  return `/api/sessions/${id}`;
};

export const updateSession = async (
  id: number,
  updateSessionBody: UpdateSessionBody,
  options?: RequestInit,
): Promise<Session> => {
  return customFetch<Session>(getUpdateSessionUrl(id), {
    ...options,
    method: "PATCH",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(updateSessionBody),
  });
};

export const getUpdateSessionMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateSession>>,
    TError,
    { id: number; data: BodyType<UpdateSessionBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof updateSession>>,
  TError,
  { id: number; data: BodyType<UpdateSessionBody> },
  TContext
> => {
  const mutationKey = ["updateSession"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof updateSession>>,
    { id: number; data: BodyType<UpdateSessionBody> }
  > = (props) => {
    const { id, data } = props ?? {};

    return updateSession(id, data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type UpdateSessionMutationResult = NonNullable<
  Awaited<ReturnType<typeof updateSession>>
>;
export type UpdateSessionMutationBody = BodyType<UpdateSessionBody>;
export type UpdateSessionMutationError = ErrorType<unknown>;

/**
 * @summary Update session
 */
export const useUpdateSession = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateSession>>,
    TError,
    { id: number; data: BodyType<UpdateSessionBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof updateSession>>,
  TError,
  { id: number; data: BodyType<UpdateSessionBody> },
  TContext
> => {
  return useMutation(getUpdateSessionMutationOptions(options));
};

/**
 * @summary List exercises
 */
export const getListExercisesUrl = (params?: ListExercisesParams) => {
  const normalizedParams = new URLSearchParams();

  Object.entries(params || {}).forEach(([key, value]) => {
    if (value !== undefined) {
      normalizedParams.append(key, value === null ? "null" : value.toString());
    }
  });

  const stringifiedParams = normalizedParams.toString();

  return stringifiedParams.length > 0
    ? `/api/exercises?${stringifiedParams}`
    : `/api/exercises`;
};

export const listExercises = async (
  params?: ListExercisesParams,
  options?: RequestInit,
): Promise<Exercise[]> => {
  return customFetch<Exercise[]>(getListExercisesUrl(params), {
    ...options,
    method: "GET",
  });
};

export const getListExercisesQueryKey = (params?: ListExercisesParams) => {
  return [`/api/exercises`, ...(params ? [params] : [])] as const;
};

export const getListExercisesQueryOptions = <
  TData = Awaited<ReturnType<typeof listExercises>>,
  TError = ErrorType<unknown>,
>(
  params?: ListExercisesParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof listExercises>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getListExercisesQueryKey(params);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof listExercises>>> = ({
    signal,
  }) => listExercises(params, { signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof listExercises>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type ListExercisesQueryResult = NonNullable<
  Awaited<ReturnType<typeof listExercises>>
>;
export type ListExercisesQueryError = ErrorType<unknown>;

/**
 * @summary List exercises
 */

export function useListExercises<
  TData = Awaited<ReturnType<typeof listExercises>>,
  TError = ErrorType<unknown>,
>(
  params?: ListExercisesParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof listExercises>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getListExercisesQueryOptions(params, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Create exercise (admin/specialist)
 */
export const getCreateExerciseUrl = () => {
  return `/api/exercises`;
};

export const createExercise = async (
  createExerciseBody: CreateExerciseBody,
  options?: RequestInit,
): Promise<Exercise> => {
  return customFetch<Exercise>(getCreateExerciseUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(createExerciseBody),
  });
};

export const getCreateExerciseMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createExercise>>,
    TError,
    { data: BodyType<CreateExerciseBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof createExercise>>,
  TError,
  { data: BodyType<CreateExerciseBody> },
  TContext
> => {
  const mutationKey = ["createExercise"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof createExercise>>,
    { data: BodyType<CreateExerciseBody> }
  > = (props) => {
    const { data } = props ?? {};

    return createExercise(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type CreateExerciseMutationResult = NonNullable<
  Awaited<ReturnType<typeof createExercise>>
>;
export type CreateExerciseMutationBody = BodyType<CreateExerciseBody>;
export type CreateExerciseMutationError = ErrorType<unknown>;

/**
 * @summary Create exercise (admin/specialist)
 */
export const useCreateExercise = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createExercise>>,
    TError,
    { data: BodyType<CreateExerciseBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof createExercise>>,
  TError,
  { data: BodyType<CreateExerciseBody> },
  TContext
> => {
  return useMutation(getCreateExerciseMutationOptions(options));
};

/**
 * @summary Get exercise by ID
 */
export const getGetExerciseUrl = (id: number) => {
  return `/api/exercises/${id}`;
};

export const getExercise = async (
  id: number,
  options?: RequestInit,
): Promise<Exercise> => {
  return customFetch<Exercise>(getGetExerciseUrl(id), {
    ...options,
    method: "GET",
  });
};

export const getGetExerciseQueryKey = (id: number) => {
  return [`/api/exercises/${id}`] as const;
};

export const getGetExerciseQueryOptions = <
  TData = Awaited<ReturnType<typeof getExercise>>,
  TError = ErrorType<unknown>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getExercise>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetExerciseQueryKey(id);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof getExercise>>> = ({
    signal,
  }) => getExercise(id, { signal, ...requestOptions });

  return {
    queryKey,
    queryFn,
    enabled: !!id,
    ...queryOptions,
  } as UseQueryOptions<
    Awaited<ReturnType<typeof getExercise>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type GetExerciseQueryResult = NonNullable<
  Awaited<ReturnType<typeof getExercise>>
>;
export type GetExerciseQueryError = ErrorType<unknown>;

/**
 * @summary Get exercise by ID
 */

export function useGetExercise<
  TData = Awaited<ReturnType<typeof getExercise>>,
  TError = ErrorType<unknown>,
>(
  id: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getExercise>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetExerciseQueryOptions(id, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Update exercise
 */
export const getUpdateExerciseUrl = (id: number) => {
  return `/api/exercises/${id}`;
};

export const updateExercise = async (
  id: number,
  updateExerciseBody: UpdateExerciseBody,
  options?: RequestInit,
): Promise<Exercise> => {
  return customFetch<Exercise>(getUpdateExerciseUrl(id), {
    ...options,
    method: "PATCH",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(updateExerciseBody),
  });
};

export const getUpdateExerciseMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateExercise>>,
    TError,
    { id: number; data: BodyType<UpdateExerciseBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof updateExercise>>,
  TError,
  { id: number; data: BodyType<UpdateExerciseBody> },
  TContext
> => {
  const mutationKey = ["updateExercise"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof updateExercise>>,
    { id: number; data: BodyType<UpdateExerciseBody> }
  > = (props) => {
    const { id, data } = props ?? {};

    return updateExercise(id, data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type UpdateExerciseMutationResult = NonNullable<
  Awaited<ReturnType<typeof updateExercise>>
>;
export type UpdateExerciseMutationBody = BodyType<UpdateExerciseBody>;
export type UpdateExerciseMutationError = ErrorType<unknown>;

/**
 * @summary Update exercise
 */
export const useUpdateExercise = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof updateExercise>>,
    TError,
    { id: number; data: BodyType<UpdateExerciseBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof updateExercise>>,
  TError,
  { id: number; data: BodyType<UpdateExerciseBody> },
  TContext
> => {
  return useMutation(getUpdateExerciseMutationOptions(options));
};

/**
 * @summary Delete exercise
 */
export const getDeleteExerciseUrl = (id: number) => {
  return `/api/exercises/${id}`;
};

export const deleteExercise = async (
  id: number,
  options?: RequestInit,
): Promise<void> => {
  return customFetch<void>(getDeleteExerciseUrl(id), {
    ...options,
    method: "DELETE",
  });
};

export const getDeleteExerciseMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof deleteExercise>>,
    TError,
    { id: number },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof deleteExercise>>,
  TError,
  { id: number },
  TContext
> => {
  const mutationKey = ["deleteExercise"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof deleteExercise>>,
    { id: number }
  > = (props) => {
    const { id } = props ?? {};

    return deleteExercise(id, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type DeleteExerciseMutationResult = NonNullable<
  Awaited<ReturnType<typeof deleteExercise>>
>;

export type DeleteExerciseMutationError = ErrorType<unknown>;

/**
 * @summary Delete exercise
 */
export const useDeleteExercise = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof deleteExercise>>,
    TError,
    { id: number },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof deleteExercise>>,
  TError,
  { id: number },
  TContext
> => {
  return useMutation(getDeleteExerciseMutationOptions(options));
};

/**
 * @summary List progress records
 */
export const getListProgressUrl = (params?: ListProgressParams) => {
  const normalizedParams = new URLSearchParams();

  Object.entries(params || {}).forEach(([key, value]) => {
    if (value !== undefined) {
      normalizedParams.append(key, value === null ? "null" : value.toString());
    }
  });

  const stringifiedParams = normalizedParams.toString();

  return stringifiedParams.length > 0
    ? `/api/progress?${stringifiedParams}`
    : `/api/progress`;
};

export const listProgress = async (
  params?: ListProgressParams,
  options?: RequestInit,
): Promise<Progress[]> => {
  return customFetch<Progress[]>(getListProgressUrl(params), {
    ...options,
    method: "GET",
  });
};

export const getListProgressQueryKey = (params?: ListProgressParams) => {
  return [`/api/progress`, ...(params ? [params] : [])] as const;
};

export const getListProgressQueryOptions = <
  TData = Awaited<ReturnType<typeof listProgress>>,
  TError = ErrorType<unknown>,
>(
  params?: ListProgressParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof listProgress>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getListProgressQueryKey(params);

  const queryFn: QueryFunction<Awaited<ReturnType<typeof listProgress>>> = ({
    signal,
  }) => listProgress(params, { signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof listProgress>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type ListProgressQueryResult = NonNullable<
  Awaited<ReturnType<typeof listProgress>>
>;
export type ListProgressQueryError = ErrorType<unknown>;

/**
 * @summary List progress records
 */

export function useListProgress<
  TData = Awaited<ReturnType<typeof listProgress>>,
  TError = ErrorType<unknown>,
>(
  params?: ListProgressParams,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof listProgress>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getListProgressQueryOptions(params, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Log progress
 */
export const getCreateProgressUrl = () => {
  return `/api/progress`;
};

export const createProgress = async (
  createProgressBody: CreateProgressBody,
  options?: RequestInit,
): Promise<Progress> => {
  return customFetch<Progress>(getCreateProgressUrl(), {
    ...options,
    method: "POST",
    headers: { "Content-Type": "application/json", ...options?.headers },
    body: JSON.stringify(createProgressBody),
  });
};

export const getCreateProgressMutationOptions = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createProgress>>,
    TError,
    { data: BodyType<CreateProgressBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationOptions<
  Awaited<ReturnType<typeof createProgress>>,
  TError,
  { data: BodyType<CreateProgressBody> },
  TContext
> => {
  const mutationKey = ["createProgress"];
  const { mutation: mutationOptions, request: requestOptions } = options
    ? options.mutation &&
      "mutationKey" in options.mutation &&
      options.mutation.mutationKey
      ? options
      : { ...options, mutation: { ...options.mutation, mutationKey } }
    : { mutation: { mutationKey }, request: undefined };

  const mutationFn: MutationFunction<
    Awaited<ReturnType<typeof createProgress>>,
    { data: BodyType<CreateProgressBody> }
  > = (props) => {
    const { data } = props ?? {};

    return createProgress(data, requestOptions);
  };

  return { mutationFn, ...mutationOptions };
};

export type CreateProgressMutationResult = NonNullable<
  Awaited<ReturnType<typeof createProgress>>
>;
export type CreateProgressMutationBody = BodyType<CreateProgressBody>;
export type CreateProgressMutationError = ErrorType<unknown>;

/**
 * @summary Log progress
 */
export const useCreateProgress = <
  TError = ErrorType<unknown>,
  TContext = unknown,
>(options?: {
  mutation?: UseMutationOptions<
    Awaited<ReturnType<typeof createProgress>>,
    TError,
    { data: BodyType<CreateProgressBody> },
    TContext
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseMutationResult<
  Awaited<ReturnType<typeof createProgress>>,
  TError,
  { data: BodyType<CreateProgressBody> },
  TContext
> => {
  return useMutation(getCreateProgressMutationOptions(options));
};

/**
 * @summary Admin dashboard summary
 */
export const getGetAdminDashboardUrl = () => {
  return `/api/dashboard/admin`;
};

export const getAdminDashboard = async (
  options?: RequestInit,
): Promise<AdminDashboard> => {
  return customFetch<AdminDashboard>(getGetAdminDashboardUrl(), {
    ...options,
    method: "GET",
  });
};

export const getGetAdminDashboardQueryKey = () => {
  return [`/api/dashboard/admin`] as const;
};

export const getGetAdminDashboardQueryOptions = <
  TData = Awaited<ReturnType<typeof getAdminDashboard>>,
  TError = ErrorType<unknown>,
>(options?: {
  query?: UseQueryOptions<
    Awaited<ReturnType<typeof getAdminDashboard>>,
    TError,
    TData
  >;
  request?: SecondParameter<typeof customFetch>;
}) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey = queryOptions?.queryKey ?? getGetAdminDashboardQueryKey();

  const queryFn: QueryFunction<
    Awaited<ReturnType<typeof getAdminDashboard>>
  > = ({ signal }) => getAdminDashboard({ signal, ...requestOptions });

  return { queryKey, queryFn, ...queryOptions } as UseQueryOptions<
    Awaited<ReturnType<typeof getAdminDashboard>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type GetAdminDashboardQueryResult = NonNullable<
  Awaited<ReturnType<typeof getAdminDashboard>>
>;
export type GetAdminDashboardQueryError = ErrorType<unknown>;

/**
 * @summary Admin dashboard summary
 */

export function useGetAdminDashboard<
  TData = Awaited<ReturnType<typeof getAdminDashboard>>,
  TError = ErrorType<unknown>,
>(options?: {
  query?: UseQueryOptions<
    Awaited<ReturnType<typeof getAdminDashboard>>,
    TError,
    TData
  >;
  request?: SecondParameter<typeof customFetch>;
}): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetAdminDashboardQueryOptions(options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Specialist dashboard summary
 */
export const getGetSpecialistDashboardUrl = (specialistId: number) => {
  return `/api/dashboard/specialist/${specialistId}`;
};

export const getSpecialistDashboard = async (
  specialistId: number,
  options?: RequestInit,
): Promise<SpecialistDashboard> => {
  return customFetch<SpecialistDashboard>(
    getGetSpecialistDashboardUrl(specialistId),
    {
      ...options,
      method: "GET",
    },
  );
};

export const getGetSpecialistDashboardQueryKey = (specialistId: number) => {
  return [`/api/dashboard/specialist/${specialistId}`] as const;
};

export const getGetSpecialistDashboardQueryOptions = <
  TData = Awaited<ReturnType<typeof getSpecialistDashboard>>,
  TError = ErrorType<unknown>,
>(
  specialistId: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getSpecialistDashboard>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey =
    queryOptions?.queryKey ?? getGetSpecialistDashboardQueryKey(specialistId);

  const queryFn: QueryFunction<
    Awaited<ReturnType<typeof getSpecialistDashboard>>
  > = ({ signal }) =>
    getSpecialistDashboard(specialistId, { signal, ...requestOptions });

  return {
    queryKey,
    queryFn,
    enabled: !!specialistId,
    ...queryOptions,
  } as UseQueryOptions<
    Awaited<ReturnType<typeof getSpecialistDashboard>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type GetSpecialistDashboardQueryResult = NonNullable<
  Awaited<ReturnType<typeof getSpecialistDashboard>>
>;
export type GetSpecialistDashboardQueryError = ErrorType<unknown>;

/**
 * @summary Specialist dashboard summary
 */

export function useGetSpecialistDashboard<
  TData = Awaited<ReturnType<typeof getSpecialistDashboard>>,
  TError = ErrorType<unknown>,
>(
  specialistId: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getSpecialistDashboard>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetSpecialistDashboardQueryOptions(
    specialistId,
    options,
  );

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Parent dashboard summary
 */
export const getGetParentDashboardUrl = (parentId: number) => {
  return `/api/dashboard/parent/${parentId}`;
};

export const getParentDashboard = async (
  parentId: number,
  options?: RequestInit,
): Promise<ParentDashboard> => {
  return customFetch<ParentDashboard>(getGetParentDashboardUrl(parentId), {
    ...options,
    method: "GET",
  });
};

export const getGetParentDashboardQueryKey = (parentId: number) => {
  return [`/api/dashboard/parent/${parentId}`] as const;
};

export const getGetParentDashboardQueryOptions = <
  TData = Awaited<ReturnType<typeof getParentDashboard>>,
  TError = ErrorType<unknown>,
>(
  parentId: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getParentDashboard>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey =
    queryOptions?.queryKey ?? getGetParentDashboardQueryKey(parentId);

  const queryFn: QueryFunction<
    Awaited<ReturnType<typeof getParentDashboard>>
  > = ({ signal }) =>
    getParentDashboard(parentId, { signal, ...requestOptions });

  return {
    queryKey,
    queryFn,
    enabled: !!parentId,
    ...queryOptions,
  } as UseQueryOptions<
    Awaited<ReturnType<typeof getParentDashboard>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type GetParentDashboardQueryResult = NonNullable<
  Awaited<ReturnType<typeof getParentDashboard>>
>;
export type GetParentDashboardQueryError = ErrorType<unknown>;

/**
 * @summary Parent dashboard summary
 */

export function useGetParentDashboard<
  TData = Awaited<ReturnType<typeof getParentDashboard>>,
  TError = ErrorType<unknown>,
>(
  parentId: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getParentDashboard>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetParentDashboardQueryOptions(parentId, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}

/**
 * @summary Child dashboard (fun data)
 */
export const getGetChildDashboardUrl = (childId: number) => {
  return `/api/dashboard/child/${childId}`;
};

export const getChildDashboard = async (
  childId: number,
  options?: RequestInit,
): Promise<ChildDashboard> => {
  return customFetch<ChildDashboard>(getGetChildDashboardUrl(childId), {
    ...options,
    method: "GET",
  });
};

export const getGetChildDashboardQueryKey = (childId: number) => {
  return [`/api/dashboard/child/${childId}`] as const;
};

export const getGetChildDashboardQueryOptions = <
  TData = Awaited<ReturnType<typeof getChildDashboard>>,
  TError = ErrorType<unknown>,
>(
  childId: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getChildDashboard>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
) => {
  const { query: queryOptions, request: requestOptions } = options ?? {};

  const queryKey =
    queryOptions?.queryKey ?? getGetChildDashboardQueryKey(childId);

  const queryFn: QueryFunction<
    Awaited<ReturnType<typeof getChildDashboard>>
  > = ({ signal }) => getChildDashboard(childId, { signal, ...requestOptions });

  return {
    queryKey,
    queryFn,
    enabled: !!childId,
    ...queryOptions,
  } as UseQueryOptions<
    Awaited<ReturnType<typeof getChildDashboard>>,
    TError,
    TData
  > & { queryKey: QueryKey };
};

export type GetChildDashboardQueryResult = NonNullable<
  Awaited<ReturnType<typeof getChildDashboard>>
>;
export type GetChildDashboardQueryError = ErrorType<unknown>;

/**
 * @summary Child dashboard (fun data)
 */

export function useGetChildDashboard<
  TData = Awaited<ReturnType<typeof getChildDashboard>>,
  TError = ErrorType<unknown>,
>(
  childId: number,
  options?: {
    query?: UseQueryOptions<
      Awaited<ReturnType<typeof getChildDashboard>>,
      TError,
      TData
    >;
    request?: SecondParameter<typeof customFetch>;
  },
): UseQueryResult<TData, TError> & { queryKey: QueryKey } {
  const queryOptions = getGetChildDashboardQueryOptions(childId, options);

  const query = useQuery(queryOptions) as UseQueryResult<TData, TError> & {
    queryKey: QueryKey;
  };

  return { ...query, queryKey: queryOptions.queryKey };
}
