// @workspace/api-zod/src/generated/api.ts
/**
 * Generated by orval v8.5.3 🍺
 * Do not edit manually.
 * Api
 * Speech Clinic API for children ages 2-5
 * OpenAPI spec version: 0.1.0
 */
import * as zod from "zod";

/**
 * @summary Health check
 */
export const HealthCheckResponse = zod.object({
  status: zod.string(),
});

/**
 * @summary Login
 */
export const LoginBody = zod.object({
  email: zod.string(),
  password: zod.string(),
});

export const LoginResponse = zod.object({
  user: zod.object({
    id: zod.number(),
    name: zod.string(),
    email: zod.string(),
    role: zod.enum(["admin", "specialist", "parent", "child"]),
    phone: zod.string().nullish(),
    createdAt: zod.string(),
  }),
  token: zod.string(),
});

/**
 * @summary Get current user
 */
export const GetCurrentUserResponse = zod.object({
  id: zod.number(),
  name: zod.string(),
  email: zod.string(),
  role: zod.enum(["admin", "specialist", "parent", "child"]),
  phone: zod.string().nullish(),
  createdAt: zod.string(),
});

/**
 * @summary List all users (admin only)
 */
export const ListUsersResponseItem = zod.object({
  id: zod.number(),
  name: zod.string(),
  email: zod.string(),
  role: zod.enum(["admin", "specialist", "parent", "child"]),
  phone: zod.string().nullish(),
  createdAt: zod.string(),
});
export const ListUsersResponse = zod.array(ListUsersResponseItem);

/**
 * @summary Create a user (admin only)
 */
export const CreateUserBody = zod.object({
  name: zod.string(),
  email: zod.string(),
  password: zod.string(),
  role: zod.enum(["admin", "specialist", "parent", "child"]),
  phone: zod.string().nullish(),
});

/**
 * @summary Get user by ID
 */
export const GetUserParams = zod.object({
  id: zod.coerce.number(),
});

export const GetUserResponse = zod.object({
  id: zod.number(),
  name: zod.string(),
  email: zod.string(),
  role: zod.enum(["admin", "specialist", "parent", "child"]),
  phone: zod.string().nullish(),
  createdAt: zod.string(),
});

/**
 * @summary Update user
 */
export const UpdateUserParams = zod.object({
  id: zod.coerce.number(),
});

export const UpdateUserBody = zod.object({
  name: zod.string().nullish(),
  phone: zod.string().nullish(),
  email: zod.string().nullish(),
});

export const UpdateUserResponse = zod.object({
  id: zod.number(),
  name: zod.string(),
  email: zod.string(),
  role: zod.enum(["admin", "specialist", "parent", "child"]),
  phone: zod.string().nullish(),
  createdAt: zod.string(),
});

/**
 * @summary Delete user
 */
export const DeleteUserParams = zod.object({
  id: zod.coerce.number(),
});

/**
 * @summary List all specialists
 */
export const ListSpecialistsResponseItem = zod.object({
  id: zod.number(),
  userId: zod.number(),
  name: zod.string(),
  email: zod.string(),
  specialty: zod.string(),
  bio: zod.string().nullish(),
  experience: zod.number().nullish(),
  rating: zod.number().nullish(),
  totalSessions: zod.number(),
  activeChildren: zod.number(),
  createdAt: zod.string(),
});
export const ListSpecialistsResponse = zod.array(ListSpecialistsResponseItem);

/**
 * @summary Create specialist profile
 */
export const CreateSpecialistBody = zod.object({
  userId: zod.number(),
  specialty: zod.string(),
  bio: zod.string().nullish(),
  experience: zod.number().nullish(),
});

/**
 * @summary Get specialist by ID
 */
export const GetSpecialistParams = zod.object({
  id: zod.coerce.number(),
});

export const GetSpecialistResponse = zod.object({
  id: zod.number(),
  userId: zod.number(),
  name: zod.string(),
  email: zod.string(),
  specialty: zod.string(),
  bio: zod.string().nullish(),
  experience: zod.number().nullish(),
  rating: zod.number().nullish(),
  totalSessions: zod.number(),
  activeChildren: zod.number(),
  createdAt: zod.string(),
});

/**
 * @summary Update specialist
 */
export const UpdateSpecialistParams = zod.object({
  id: zod.coerce.number(),
});

export const UpdateSpecialistBody = zod.object({
  specialty: zod.string().nullish(),
  bio: zod.string().nullish(),
  experience: zod.number().nullish(),
});

export const UpdateSpecialistResponse = zod.object({
  id: zod.number(),
  userId: zod.number(),
  name: zod.string(),
  email: zod.string(),
  specialty: zod.string(),
  bio: zod.string().nullish(),
  experience: zod.number().nullish(),
  rating: zod.number().nullish(),
  totalSessions: zod.number(),
  activeChildren: zod.number(),
  createdAt: zod.string(),
});

/**
 * @summary List children
 */
export const ListChildrenQueryParams = zod.object({
  parentId: zod.coerce.number().nullish(),
  specialistId: zod.coerce.number().nullish(),
});

export const ListChildrenResponseItem = zod.object({
  id: zod.number(),
  name: zod.string(),
  age: zod.number(),
  gender: zod.enum(["male", "female"]),
  parentId: zod.number(),
  specialistId: zod.number().nullish(),
  diagnosis: zod.string().nullish(),
  notes: zod.string().nullish(),
  progressLevel: zod.number(),
  totalSessions: zod.number(),
  createdAt: zod.string(),
});
export const ListChildrenResponse = zod.array(ListChildrenResponseItem);

/**
 * @summary Create child profile
 */
export const CreateChildBody = zod.object({
  name: zod.string(),
  age: zod.number(),
  gender: zod.enum(["male", "female"]),
  parentId: zod.number(),
  specialistId: zod.number().nullish(),
  diagnosis: zod.string().nullish(),
  notes: zod.string().nullish(),
});

/**
 * @summary Get child by ID
 */
export const GetChildParams = zod.object({
  id: zod.coerce.number(),
});

export const GetChildResponse = zod.object({
  id: zod.number(),
  name: zod.string(),
  age: zod.number(),
  gender: zod.enum(["male", "female"]),
  parentId: zod.number(),
  specialistId: zod.number().nullish(),
  diagnosis: zod.string().nullish(),
  notes: zod.string().nullish(),
  progressLevel: zod.number(),
  totalSessions: zod.number(),
  createdAt: zod.string(),
});

/**
 * @summary Update child
 */
export const UpdateChildParams = zod.object({
  id: zod.coerce.number(),
});

export const UpdateChildBody = zod.object({
  name: zod.string().nullish(),
  age: zod.number().nullish(),
  specialistId: zod.number().nullish(),
  diagnosis: zod.string().nullish(),
  notes: zod.string().nullish(),
  progressLevel: zod.number().nullish(),
});

export const UpdateChildResponse = zod.object({
  id: zod.number(),
  name: zod.string(),
  age: zod.number(),
  gender: zod.enum(["male", "female"]),
  parentId: zod.number(),
  specialistId: zod.number().nullish(),
  diagnosis: zod.string().nullish(),
  notes: zod.string().nullish(),
  progressLevel: zod.number(),
  totalSessions: zod.number(),
  createdAt: zod.string(),
});

/**
 * @summary Delete child
 */
export const DeleteChildParams = zod.object({
  id: zod.coerce.number(),
});

/**
 * @summary List appointments
 */
export const ListAppointmentsQueryParams = zod.object({
  childId: zod.coerce.number().nullish(),
  specialistId: zod.coerce.number().nullish(),
  status: zod.coerce.string().nullish(),
});

export const ListAppointmentsResponseItem = zod.object({
  id: zod.number(),
  childId: zod.number(),
  childName: zod.string(),
  specialistId: zod.number(),
  specialistName: zod.string(),
  date: zod.string(),
  time: zod.string(),
  duration: zod.number(),
  status: zod.enum(["scheduled", "completed", "cancelled", "no_show"]),
  notes: zod.string().nullish(),
  createdAt: zod.string(),
});
export const ListAppointmentsResponse = zod.array(ListAppointmentsResponseItem);

/**
 * @summary Create appointment
 */
export const CreateAppointmentBody = zod.object({
  childId: zod.number(),
  specialistId: zod.number(),
  date: zod.string(),
  time: zod.string(),
  duration: zod.number(),
  notes: zod.string().nullish(),
});

/**
 * @summary Get appointment by ID
 */
export const GetAppointmentParams = zod.object({
  id: zod.coerce.number(),
});

export const GetAppointmentResponse = zod.object({
  id: zod.number(),
  childId: zod.number(),
  childName: zod.string(),
  specialistId: zod.number(),
  specialistName: zod.string(),
  date: zod.string(),
  time: zod.string(),
  duration: zod.number(),
  status: zod.enum(["scheduled", "completed", "cancelled", "no_show"]),
  notes: zod.string().nullish(),
  createdAt: zod.string(),
});

/**
 * @summary Update appointment
 */
export const UpdateAppointmentParams = zod.object({
  id: zod.coerce.number(),
});

export const UpdateAppointmentBody = zod.object({
  date: zod.string().nullish(),
  time: zod.string().nullish(),
  duration: zod.number().nullish(),
  status: zod.string().nullish(),
  notes: zod.string().nullish(),
});

export const UpdateAppointmentResponse = zod.object({
  id: zod.number(),
  childId: zod.number(),
  childName: zod.string(),
  specialistId: zod.number(),
  specialistName: zod.string(),
  date: zod.string(),
  time: zod.string(),
  duration: zod.number(),
  status: zod.enum(["scheduled", "completed", "cancelled", "no_show"]),
  notes: zod.string().nullish(),
  createdAt: zod.string(),
});

/**
 * @summary Delete appointment
 */
export const DeleteAppointmentParams = zod.object({
  id: zod.coerce.number(),
});

/**
 * @summary List therapy sessions
 */
export const ListSessionsQueryParams = zod.object({
  childId: zod.coerce.number().nullish(),
  specialistId: zod.coerce.number().nullish(),
});

export const ListSessionsResponseItem = zod.object({
  id: zod.number(),
  childId: zod.number(),
  childName: zod.string(),
  specialistId: zod.number(),
  specialistName: zod.string(),
  date: zod.string(),
  duration: zod.number(),
  score: zod.number().nullish(),
  notes: zod.string().nullish(),
  exercisesCompleted: zod.number(),
  mood: zod.string().nullish(),
  createdAt: zod.string(),
});
export const ListSessionsResponse = zod.array(ListSessionsResponseItem);

/**
 * @summary Create therapy session
 */
export const CreateSessionBody = zod.object({
  childId: zod.number(),
  specialistId: zod.number(),
  date: zod.string(),
  duration: zod.number(),
  score: zod.number().nullish(),
  notes: zod.string().nullish(),
  exercisesCompleted: zod.number(),
  mood: zod.string().nullish(),
});

/**
 * @summary Get session by ID
 */
export const GetSessionParams = zod.object({
  id: zod.coerce.number(),
});

export const GetSessionResponse = zod.object({
  id: zod.number(),
  childId: zod.number(),
  childName: zod.string(),
  specialistId: zod.number(),
  specialistName: zod.string(),
  date: zod.string(),
  duration: zod.number(),
  score: zod.number().nullish(),
  notes: zod.string().nullish(),
  exercisesCompleted: zod.number(),
  mood: zod.string().nullish(),
  createdAt: zod.string(),
});

/**
 * @summary Update session
 */
export const UpdateSessionParams = zod.object({
  id: zod.coerce.number(),
});

export const UpdateSessionBody = zod.object({
  score: zod.number().nullish(),
  notes: zod.string().nullish(),
  exercisesCompleted: zod.number().nullish(),
  mood: zod.string().nullish(),
});

export const UpdateSessionResponse = zod.object({
  id: zod.number(),
  childId: zod.number(),
  childName: zod.string(),
  specialistId: zod.number(),
  specialistName: zod.string(),
  date: zod.string(),
  duration: zod.number(),
  score: zod.number().nullish(),
  notes: zod.string().nullish(),
  exercisesCompleted: zod.number(),
  mood: zod.string().nullish(),
  createdAt: zod.string(),
});

/**
 * @summary List exercises
 */
export const ListExercisesQueryParams = zod.object({
  category: zod.coerce.string().nullish(),
  difficulty: zod.coerce.string().nullish(),
  minAge: zod.coerce.number().nullish(),
  maxAge: zod.coerce.number().nullish(),
});

export const ListExercisesResponseItem = zod.object({
  id: zod.number(),
  title: zod.string(),
  description: zod.string(),
  category: zod.enum([
    "pronunciation",
    "vocabulary",
    "fluency",
    "listening",
    "breathing",
  ]),
  difficulty: zod.enum(["easy", "medium", "hard"]),
  minAge: zod.number(),
  maxAge: zod.number(),
  instructions: zod.string(),
  duration: zod.number(),
  emoji: zod.string(),
  createdAt: zod.string(),
});
export const ListExercisesResponse = zod.array(ListExercisesResponseItem);

/**
 * @summary Create exercise (admin/specialist)
 */
export const CreateExerciseBody = zod.object({
  title: zod.string(),
  description: zod.string(),
  category: zod.enum([
    "pronunciation",
    "vocabulary",
    "fluency",
    "listening",
    "breathing",
  ]),
  difficulty: zod.enum(["easy", "medium", "hard"]),
  minAge: zod.number(),
  maxAge: zod.number(),
  instructions: zod.string(),
  duration: zod.number(),
  emoji: zod.string(),
});

/**
 * @summary Get exercise by ID
 */
export const GetExerciseParams = zod.object({
  id: zod.coerce.number(),
});

export const GetExerciseResponse = zod.object({
  id: zod.number(),
  title: zod.string(),
  description: zod.string(),
  category: zod.enum([
    "pronunciation",
    "vocabulary",
    "fluency",
    "listening",
    "breathing",
  ]),
  difficulty: zod.enum(["easy", "medium", "hard"]),
  minAge: zod.number(),
  maxAge: zod.number(),
  instructions: zod.string(),
  duration: zod.number(),
  emoji: zod.string(),
  createdAt: zod.string(),
});

/**
 * @summary Update exercise
 */
export const UpdateExerciseParams = zod.object({
  id: zod.coerce.number(),
});

export const UpdateExerciseBody = zod.object({
  title: zod.string().nullish(),
  description: zod.string().nullish(),
  category: zod.string().nullish(),
  difficulty: zod.string().nullish(),
  instructions: zod.string().nullish(),
  duration: zod.number().nullish(),
});

export const UpdateExerciseResponse = zod.object({
  id: zod.number(),
  title: zod.string(),
  description: zod.string(),
  category: zod.enum([
    "pronunciation",
    "vocabulary",
    "fluency",
    "listening",
    "breathing",
  ]),
  difficulty: zod.enum(["easy", "medium", "hard"]),
  minAge: zod.number(),
  maxAge: zod.number(),
  instructions: zod.string(),
  duration: zod.number(),
  emoji: zod.string(),
  createdAt: zod.string(),
});

/**
 * @summary Delete exercise
 */
export const DeleteExerciseParams = zod.object({
  id: zod.coerce.number(),
});

/**
 * @summary List progress records
 */
export const ListProgressQueryParams = zod.object({
  childId: zod.coerce.number().nullish(),
  exerciseId: zod.coerce.number().nullish(),
});

export const ListProgressResponseItem = zod.object({
  id: zod.number(),
  childId: zod.number(),
  exerciseId: zod.number(),
  exerciseTitle: zod.string(),
  score: zod.number(),
  completed: zod.boolean(),
  date: zod.string(),
  notes: zod.string().nullish(),
  createdAt: zod.string(),
});
export const ListProgressResponse = zod.array(ListProgressResponseItem);

/**
 * @summary Log progress
 */
export const CreateProgressBody = zod.object({
  childId: zod.number(),
  exerciseId: zod.number(),
  score: zod.number(),
  completed: zod.boolean(),
  date: zod.string(),
  notes: zod.string().nullish(),
});

/**
 * @summary Admin dashboard summary
 */
export const GetAdminDashboardResponse = zod.object({
  totalChildren: zod.number(),
  totalSpecialists: zod.number(),
  totalParents: zod.number(),
  totalSessions: zod.number(),
  appointmentsToday: zod.number(),
  appointmentsThisWeek: zod.number(),
  averageProgress: zod.number(),
  recentSessions: zod.array(
    zod.object({
      id: zod.number(),
      childId: zod.number(),
      childName: zod.string(),
      specialistId: zod.number(),
      specialistName: zod.string(),
      date: zod.string(),
      duration: zod.number(),
      score: zod.number().nullish(),
      notes: zod.string().nullish(),
      exercisesCompleted: zod.number(),
      mood: zod.string().nullish(),
      createdAt: zod.string(),
    }),
  ),
  upcomingAppointments: zod.array(
    zod.object({
      id: zod.number(),
      childId: zod.number(),
      childName: zod.string(),
      specialistId: zod.number(),
      specialistName: zod.string(),
      date: zod.string(),
      time: zod.string(),
      duration: zod.number(),
      status: zod.enum(["scheduled", "completed", "cancelled", "no_show"]),
      notes: zod.string().nullish(),
      createdAt: zod.string(),
    }),
  ),
});

/**
 * @summary Specialist dashboard summary
 */
export const GetSpecialistDashboardParams = zod.object({
  specialistId: zod.coerce.number(),
});

export const GetSpecialistDashboardResponse = zod.object({
  activeChildren: zod.number(),
  sessionsThisWeek: zod.number(),
  sessionsThisMonth: zod.number(),
  averageScore: zod.number(),
  upcomingAppointments: zod.array(
    zod.object({
      id: zod.number(),
      childId: zod.number(),
      childName: zod.string(),
      specialistId: zod.number(),
      specialistName: zod.string(),
      date: zod.string(),
      time: zod.string(),
      duration: zod.number(),
      status: zod.enum(["scheduled", "completed", "cancelled", "no_show"]),
      notes: zod.string().nullish(),
      createdAt: zod.string(),
    }),
  ),
  recentSessions: zod.array(
    zod.object({
      id: zod.number(),
      childId: zod.number(),
      childName: zod.string(),
      specialistId: zod.number(),
      specialistName: zod.string(),
      date: zod.string(),
      duration: zod.number(),
      score: zod.number().nullish(),
      notes: zod.string().nullish(),
      exercisesCompleted: zod.number(),
      mood: zod.string().nullish(),
      createdAt: zod.string(),
    }),
  ),
  children: zod.array(
    zod.object({
      id: zod.number(),
      name: zod.string(),
      age: zod.number(),
      gender: zod.enum(["male", "female"]),
      parentId: zod.number(),
      specialistId: zod.number().nullish(),
      diagnosis: zod.string().nullish(),
      notes: zod.string().nullish(),
      progressLevel: zod.number(),
      totalSessions: zod.number(),
      createdAt: zod.string(),
    }),
  ),
});

/**
 * @summary Parent dashboard summary
 */
export const GetParentDashboardParams = zod.object({
  parentId: zod.coerce.number(),
});

export const GetParentDashboardResponse = zod.object({
  children: zod.array(
    zod.object({
      id: zod.number(),
      name: zod.string(),
      age: zod.number(),
      gender: zod.enum(["male", "female"]),
      parentId: zod.number(),
      specialistId: zod.number().nullish(),
      diagnosis: zod.string().nullish(),
      notes: zod.string().nullish(),
      progressLevel: zod.number(),
      totalSessions: zod.number(),
      createdAt: zod.string(),
    }),
  ),
  upcomingAppointments: zod.array(
    zod.object({
      id: zod.number(),
      childId: zod.number(),
      childName: zod.string(),
      specialistId: zod.number(),
      specialistName: zod.string(),
      date: zod.string(),
      time: zod.string(),
      duration: zod.number(),
      status: zod.enum(["scheduled", "completed", "cancelled", "no_show"]),
      notes: zod.string().nullish(),
      createdAt: zod.string(),
    }),
  ),
  recentSessions: zod.array(
    zod.object({
      id: zod.number(),
      childId: zod.number(),
      childName: zod.string(),
      specialistId: zod.number(),
      specialistName: zod.string(),
      date: zod.string(),
      duration: zod.number(),
      score: zod.number().nullish(),
      notes: zod.string().nullish(),
      exercisesCompleted: zod.number(),
      mood: zod.string().nullish(),
      createdAt: zod.string(),
    }),
  ),
  totalProgress: zod.number(),
  lastSessionDate: zod.string().nullish(),
});

/**
 * @summary Child dashboard (fun data)
 */
export const GetChildDashboardParams = zod.object({
  childId: zod.coerce.number(),
});

export const GetChildDashboardResponse = zod.object({
  childName: zod.string(),
  stars: zod.number(),
  level: zod.number(),
  todayExercises: zod.array(
    zod.object({
      id: zod.number(),
      title: zod.string(),
      description: zod.string(),
      category: zod.enum([
        "pronunciation",
        "vocabulary",
        "fluency",
        "listening",
        "breathing",
      ]),
      difficulty: zod.enum(["easy", "medium", "hard"]),
      minAge: zod.number(),
      maxAge: zod.number(),
      instructions: zod.string(),
      duration: zod.number(),
      emoji: zod.string(),
      createdAt: zod.string(),
    }),
  ),
  recentProgress: zod.array(
    zod.object({
      id: zod.number(),
      childId: zod.number(),
      exerciseId: zod.number(),
      exerciseTitle: zod.string(),
      score: zod.number(),
      completed: zod.boolean(),
      date: zod.string(),
      notes: zod.string().nullish(),
      createdAt: zod.string(),
    }),
  ),
  nextAppointment: zod.string().nullish(),
  streakDays: zod.number(),
});
