// @workspace/api-client-react/src/generated/api.schemas.ts
/**
 * Generated by orval v8.5.3 🍺
 * Do not edit manually.
 * Api
 * Speech Clinic API for children ages 2-5
 * OpenAPI spec version: 0.1.0
 */
export interface HealthStatus {
  status: string;
}

export interface LoginBody {
  email: string;
  password: string;
}

export type UserRole = (typeof UserRole)[keyof typeof UserRole];

export const UserRole = {
  admin: "admin",
  specialist: "specialist",
  parent: "parent",
  child: "child",
} as const;

export interface User {
  id: number;
  name: string;
  email: string;
  role: UserRole;
  /** @nullable */
  phone?: string | null;
  createdAt: string;
}

export interface LoginResponse {
  user: User;
  token: string;
}

export type CreateUserBodyRole =
  (typeof CreateUserBodyRole)[keyof typeof CreateUserBodyRole];

export const CreateUserBodyRole = {
  admin: "admin",
  specialist: "specialist",
  parent: "parent",
  child: "child",
} as const;

export interface CreateUserBody {
  name: string;
  email: string;
  password: string;
  role: CreateUserBodyRole;
  /** @nullable */
  phone?: string | null;
}

export interface UpdateUserBody {
  /** @nullable */
  name?: string | null;
  /** @nullable */
  phone?: string | null;
  /** @nullable */
  email?: string | null;
}

export interface Specialist {
  id: number;
  userId: number;
  name: string;
  email: string;
  specialty: string;
  /** @nullable */
  bio?: string | null;
  /** @nullable */
  experience?: number | null;
  /** @nullable */
  rating?: number | null;
  totalSessions: number;
  activeChildren: number;
  createdAt: string;
}

export interface CreateSpecialistBody {
  userId: number;
  specialty: string;
  /** @nullable */
  bio?: string | null;
  /** @nullable */
  experience?: number | null;
}

export interface UpdateSpecialistBody {
  /** @nullable */
  specialty?: string | null;
  /** @nullable */
  bio?: string | null;
  /** @nullable */
  experience?: number | null;
}

export type ChildGender = (typeof ChildGender)[keyof typeof ChildGender];

export const ChildGender = {
  male: "male",
  female: "female",
} as const;

export interface Child {
  id: number;
  name: string;
  age: number;
  gender: ChildGender;
  parentId: number;
  /** @nullable */
  specialistId?: number | null;
  /** @nullable */
  diagnosis?: string | null;
  /** @nullable */
  notes?: string | null;
  progressLevel: number;
  totalSessions: number;
  createdAt: string;
}

export type CreateChildBodyGender =
  (typeof CreateChildBodyGender)[keyof typeof CreateChildBodyGender];

export const CreateChildBodyGender = {
  male: "male",
  female: "female",
} as const;

export interface CreateChildBody {
  name: string;
  age: number;
  gender: CreateChildBodyGender;
  parentId: number;
  /** @nullable */
  specialistId?: number | null;
  /** @nullable */
  diagnosis?: string | null;
  /** @nullable */
  notes?: string | null;
}

export interface UpdateChildBody {
  /** @nullable */
  name?: string | null;
  /** @nullable */
  age?: number | null;
  /** @nullable */
  specialistId?: number | null;
  /** @nullable */
  diagnosis?: string | null;
  /** @nullable */
  notes?: string | null;
  /** @nullable */
  progressLevel?: number | null;
}

export type AppointmentStatus =
  (typeof AppointmentStatus)[keyof typeof AppointmentStatus];

export const AppointmentStatus = {
  scheduled: "scheduled",
  completed: "completed",
  cancelled: "cancelled",
  no_show: "no_show",
} as const;

export interface Appointment {
  id: number;
  childId: number;
  childName: string;
  specialistId: number;
  specialistName: string;
  date: string;
  time: string;
  duration: number;
  status: AppointmentStatus;
  /** @nullable */
  notes?: string | null;
  createdAt: string;
}

export interface CreateAppointmentBody {
  childId: number;
  specialistId: number;
  date: string;
  time: string;
  duration: number;
  /** @nullable */
  notes?: string | null;
}

export interface UpdateAppointmentBody {
  /** @nullable */
  date?: string | null;
  /** @nullable */
  time?: string | null;
  /** @nullable */
  duration?: number | null;
  /** @nullable */
  status?: string | null;
  /** @nullable */
  notes?: string | null;
}

export interface Session {
  id: number;
  childId: number;
  childName: string;
  specialistId: number;
  specialistName: string;
  date: string;
  duration: number;
  /** @nullable */
  score?: number | null;
  /** @nullable */
  notes?: string | null;
  exercisesCompleted: number;
  /** @nullable */
  mood?: string | null;
  createdAt: string;
}

export interface CreateSessionBody {
  childId: number;
  specialistId: number;
  date: string;
  duration: number;
  /** @nullable */
  score?: number | null;
  /** @nullable */
  notes?: string | null;
  exercisesCompleted: number;
  /** @nullable */
  mood?: string | null;
}

export interface UpdateSessionBody {
  /** @nullable */
  score?: number | null;
  /** @nullable */
  notes?: string | null;
  /** @nullable */
  exercisesCompleted?: number | null;
  /** @nullable */
  mood?: string | null;
}

export type ExerciseCategory =
  (typeof ExerciseCategory)[keyof typeof ExerciseCategory];

export const ExerciseCategory = {
  pronunciation: "pronunciation",
  vocabulary: "vocabulary",
  fluency: "fluency",
  listening: "listening",
  breathing: "breathing",
} as const;

export type ExerciseDifficulty =
  (typeof ExerciseDifficulty)[keyof typeof ExerciseDifficulty];

export const ExerciseDifficulty = {
  easy: "easy",
  medium: "medium",
  hard: "hard",
} as const;

export interface Exercise {
  id: number;
  title: string;
  description: string;
  category: ExerciseCategory;
  difficulty: ExerciseDifficulty;
  minAge: number;
  maxAge: number;
  instructions: string;
  duration: number;
  emoji: string;
  createdAt: string;
}

export type CreateExerciseBodyCategory =
  (typeof CreateExerciseBodyCategory)[keyof typeof CreateExerciseBodyCategory];

export const CreateExerciseBodyCategory = {
  pronunciation: "pronunciation",
  vocabulary: "vocabulary",
  fluency: "fluency",
  listening: "listening",
  breathing: "breathing",
} as const;

export type CreateExerciseBodyDifficulty =
  (typeof CreateExerciseBodyDifficulty)[keyof typeof CreateExerciseBodyDifficulty];

export const CreateExerciseBodyDifficulty = {
  easy: "easy",
  medium: "medium",
  hard: "hard",
} as const;

export interface CreateExerciseBody {
  title: string;
  description: string;
  category: CreateExerciseBodyCategory;
  difficulty: CreateExerciseBodyDifficulty;
  minAge: number;
  maxAge: number;
  instructions: string;
  duration: number;
  emoji: string;
}

export interface UpdateExerciseBody {
  /** @nullable */
  title?: string | null;
  /** @nullable */
  description?: string | null;
  /** @nullable */
  category?: string | null;
  /** @nullable */
  difficulty?: string | null;
  /** @nullable */
  instructions?: string | null;
  /** @nullable */
  duration?: number | null;
}

export interface Progress {
  id: number;
  childId: number;
  exerciseId: number;
  exerciseTitle: string;
  score: number;
  completed: boolean;
  date: string;
  /** @nullable */
  notes?: string | null;
  createdAt: string;
}

export interface CreateProgressBody {
  childId: number;
  exerciseId: number;
  score: number;
  completed: boolean;
  date: string;
  /** @nullable */
  notes?: string | null;
}

export interface AdminDashboard {
  totalChildren: number;
  totalSpecialists: number;
  totalParents: number;
  totalSessions: number;
  appointmentsToday: number;
  appointmentsThisWeek: number;
  averageProgress: number;
  recentSessions: Session[];
  upcomingAppointments: Appointment[];
}

export interface SpecialistDashboard {
  activeChildren: number;
  sessionsThisWeek: number;
  sessionsThisMonth: number;
  averageScore: number;
  upcomingAppointments: Appointment[];
  recentSessions: Session[];
  children: Child[];
}

export interface ParentDashboard {
  children: Child[];
  upcomingAppointments: Appointment[];
  recentSessions: Session[];
  totalProgress: number;
  /** @nullable */
  lastSessionDate?: string | null;
}

export interface ChildDashboard {
  childName: string;
  stars: number;
  level: number;
  todayExercises: Exercise[];
  recentProgress: Progress[];
  /** @nullable */
  nextAppointment?: string | null;
  streakDays: number;
}

export type ListChildrenParams = {
  /**
   * @nullable
   */
  parentId?: number | null;
  /**
   * @nullable
   */
  specialistId?: number | null;
};

export type ListAppointmentsParams = {
  /**
   * @nullable
   */
  childId?: number | null;
  /**
   * @nullable
   */
  specialistId?: number | null;
  /**
   * @nullable
   */
  status?: string | null;
};

export type ListSessionsParams = {
  /**
   * @nullable
   */
  childId?: number | null;
  /**
   * @nullable
   */
  specialistId?: number | null;
};

export type ListExercisesParams = {
  /**
   * @nullable
   */
  category?: string | null;
  /**
   * @nullable
   */
  difficulty?: string | null;
  /**
   * @nullable
   */
  minAge?: number | null;
  /**
   * @nullable
   */
  maxAge?: number | null;
};

export type ListProgressParams = {
  /**
   * @nullable
   */
  childId?: number | null;
  /**
   * @nullable
   */
  exerciseId?: number | null;
};
